design/configuration-crd.md

```diff
@@ -0,0 +1,210 @@
+
+## Contour Deployment
+
+A managed version of Contour was made available with the `Contour Operator`.
+Since Contour will manage Envoy instances, the Operator will now manage instances of Contour.
+The details of how an instance of Contour should be deployed within a cluster will be defined in the second CRD named `ContourDeployment`. 
+The `spec.confguration` of this object will be the same struct defined in the `ContourConfiguration`. 
+
+A controller will watch for these objects to be created and take action on them accordingly to make desired state in the cluster match the configuration on the spec. 
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ContourDeployment
+metadata:
+  name: contour
+spec:
+  replicas: 2
+  nodePlacement:
+    nodeSelector:
+    tolerations:
+  configuration:
+    <same config as above>
+status:
+```
+
+## Processing Logic
+
+Contour will require a new flag (`--contour-config`), which will allow for customizing the name of the `ContourConfiguration` CRD that is it to use.
+It will default to one named `contour`, but could be overridden if desired.
+The ContourConfiguration referenced must also be in the same namespace as Contour is running, it's not valid to reference a configuration from another namespace.
+The current flag `--config-path`/`-c` will continue to point to the Configmap file, but over time could eventually be deprecated and the short code `-c` be used for the CRD location (i.e. `--contour-config`) for simplicity.
+The Contour configuration CRD will still remain optional.
+In its absence, Contour will operate with reasonable defaults.
+Where Contour settings can also be specified with command-line flags, the command-line value takes precedence over the configuration file.
+
+On startup, Contour will look for a `ContourConfiguration` CRD named `contour` or using the value of the flag `--contour-config` in the same namespace which Contour is running in, Contour won't support referencing from a different namespace.
+If the `ContourConfiguration` CRD is not found Contour will start up with reasonable defaults.
+
+Contour will set status on the object informing the user if there are any errors or issues with the config file or that is all fine and processing correctly.
+
+Once Contour begins using a Configuration CRD, it will add a finalizer to it such that if that resource is going to get deleted, Contour is aware of it.
+Should the Configuration CRD be deleted while it is in use, Contour will default back to reasonable defaults and log the issue.
+
+When config in the CRD changes we will gracefully stop the dependent ingress/gateway controllers and restart them with new config, or dynamically update some in-memory data that the controllers use.
+
+If the configuration changes, Contour will first validate the new Configuration.
+If that new change set results in the object being invalid, Contour won't restart its controllers but set status and still try and serve the old configuration.
```

Comment:
The downside to this approach is that if config becomes invalid, then there's no real record of what config Contour *is* running with since what's now in the CRD spec does not match what Contour is using. If we do want to implement this logic, then I think we should have some place where the "active" Config exists in a CRD. Could have a separate CRD representing a "revision" of ContourConfig, or just write the last good config out in `ContourConfig` itself, or something else..